### PR TITLE
refactor(rolldown): simplify the runtime module creation

### PR DESCRIPTION
**Pull request by Brooooooklyn**
### Description

Avoid to create an async task for runtime module creation.

### CODE REVIEW

**Comment by Brooooooklyn** [comment]
* **#874** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/874?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Brooooooklyn and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by hyf0** [comment]
Purpose of original code is we don't want parsing runtime module to block the main thread logic. We could simplify them, but what's the goal here?

---
edited

We also don't think through whether to expose the runtime module via hooks. However, if we gonna do this way, it means we might need execture js code here. If we make it sync, it will just block the whole bundling process.

**Comment by Brooooooklyn** [comment]
> Purpose of original code is we don't want parsing runtime module to block the main thread logic.

The `ModuleLoader::new` is already in a nested async call stack, which means It doesn't necessarily happen in the main thread. Spawning merely moves computation from one tokio worker to another tokio worker, which I believe is an unnecessary overhead.

Regarding the optimization of Tokio scheduling, I think it can wait until there is a large real-world project that can be connected, and then fine-tune through [tokio_console](https://github.com/tokio-rs/console). Before that, let's prioritize simplifying the code complexity and making the code control flow clearer.

**Comment by hyf0** [comment]
> > Purpose of original code is we don't want parsing runtime module to block the main thread logic.
> 
> The `ModuleLoader::new` is already in a nested async call stack, which means It doesn't necessarily happen in the main thread. Spawning merely moves computation from one tokio worker to another tokio worker, which I believe is an unnecessary overhead.
> 
> Regarding the optimization of Tokio scheduling, I think it can wait until there is a large real-world project that can be connected, and then fine-tune through [tokio_console](https://github.com/tokio-rs/console). Before that, let's prioritize simplifying the code complexity and making the code control flow clearer.

Changes you made here https://github.com/rolldown/rolldown/pull/874/files#diff-79e9098941ee0e7d5a23b2dba8b426dd3b85aafcad407c9606f3788c2858d819L56-L64 means that creating and scanning runtime module will definitely run after `resolve_user_defined_entries` while the previous is running in parallel if tokio workers is running in multiple thread mode.

Tokio runtime will do the best effort while it can with the original code. However, with this PR changed, it would never be running in parallel.

**Comment by Brooooooklyn** [comment]
I tested the execution time of create_runtime_module, and it's about **560.917µs** on my computer.
image

I feel it's not worth it to schedule an extra task for this level of computational workload. 

And I found that `RuntimeNormalModuleTaskResult` is almost a constant; maybe we can make it static in the future (by computing in advance during compilation through `build.rs`). Before this, I think it is worthwhile to simplify the logic and program control flow used to create it.

**Comment by hyf0** [comment]
> And I found that `RuntimeNormalModuleTaskResult` is almost a constant; maybe we can make it static in the future

Yeah. Runtime module needs to be generated dynamically depends on user configuration. If user want to generated code targeting es5 and es6 or others.

https://rollupjs.org/configuration-options/#output-generatedcode

~Notice that this option allows to accept a function. As mush as I don't want to support accepting a function, but~ all of that show that the complexity to deal with runtime module is enough to encapsulate it in a single task and do everything in it.

---

The principle here is always to run parallel if possible because we want to as soon as possible to warm up the parsing process of user's code. This part is the key step of bundling, simplicity isn't over performance here. I also know that current execution time of runtime module task is very short, but we still spawned it. Because we know we would add more code here.

We used to parse runtime module in sync and refactor it to make it ran in async due to the above reasons. However, if we already know what would happen most likely, guess it would be good to make the code more satisfy the future

Related issues:
- https://github.com/rolldown/rolldown/issues/206

**Comment by hyf0** [comment]
https://github.com/evanw/esbuild/blob/7a225fff7233a8de275c016e2215a6f9ea356030/internal/bundler/bundler.go#L1283 does it here. I'm not saying what esbuild does is definitely correct. But in this case, it doesn't show much negative, so it could be sign to infer us how the code is gonna like to be in the future.

---edited

I just found that esbuild does this more early. It parses runtime module even before call `build_start` hook.

**Comment by Brooooooklyn** [comment]
@hyf0 made slight modifications to optimize the program flow, keeping the logic for task run spawn.

File: crates/rolldown/src/module_loader/module_loader.rs [hunk 1]
```diff
@@ -82,16 +84,38 @@ impl ModuleLoader {
...
+    let mut intermediate_normal_modules = IntermediateNormalModules::new();
+    let mut symbols = Symbols::default();
+    let runtime_id = intermediate_normal_modules.alloc_module_id(&mut symbols);
+
+    let task = RuntimeNormalModuleTask::new(runtime_id, tx_to_runtime_module);
+    let handle = tokio::runtime::Handle::current();
+
+    #[cfg(target_family = "wasm")]
+    {
+      // could not block_on/spawn the main thread in WASI
```
Comment (hyf0) [inline crates/rolldown/src/module_loader/module_loader.rs:96]: Changes are fine. But this comment seems a little confused with `could not block_on` and the following `handle.block_on(`.

**Review by hyf0 — APPROVED** [review]

**Pull request merged**